lib/configuration.js

```diff
@@ -15,150 +15,430 @@
+   * number identifier for the project. If the Configuration instance is unable
+   * to retrieve a project number from the metadata service or the runtime
+   * configuration then the property will remain null. If given both a project
+   * number through the metadata service and the runtime configuration then the
+   * instance will assign the value given by the metadata service over the
+   * runtime configuration. If the Configuration instance is not able to 
+   * retrieve a project number or a project id from runtime configuration or
+   * from the metadata service then this will trigger the `error` event in which
+   * listening components must operate in 'offline' mode.
+   * {@link https://cloud.google.com/compute/docs/storing-retrieving-metadata}
+   * @memberof Configuration
+   * @private
+   * @type {String|Null}
+   * @defaultvalue null
+   */
+  this._projectNumber = null;
+  /**
+   * The _key property is meant to contain the optional Stackdriver API key that
+   * may be used in place of default application credentials to authenticate
+   * with the Stackdriver Error API. This property will remain null if a key
+   * is not given in the runtime configuration or an invalid type is given as
+   * the runtime configuration.
+   * {@link https://support.google.com/cloud/answer/6158862?hl=en}
+   * @memberof Configuration
+   * @private
+   * @type {String|Null}
+   * @defaultvalue null
+   */
+  this._key = null;
+  /**
+   * The _serviceContext property is meant to contain the optional service
+   * context information which may be given in the runtime configuration. If
+   * not given in the runtime configuration then the property value will remain
+   * null.
+   * @memberof Configuration
+   * @private
+   * @type {Object}
+   * @default
+   */
+  this._serviceContext = {service: '', version: ''};
+  /**
+   * The _version of the Error reporting library that is currently being run.
+   * This information will be logged in errors communicated to the Stackdriver
+   * Error API.
+   * @memberof Configuration
+   * @private
+   * @type {String}
+   */
+  this._version = version;
+  /**
+   * The _givenConfiguration property holds a ConfigurationOptions object
+   * which, if valid, will be merged against by the values taken from the meta-
+   * data service. If the _givenConfiguration property is not valid then only
+   * metadata values will be used in the Configuration instance.
+   * @memberof Configuration
+   * @private
+   * @type {Object|Null}
+   * @defaultvalue null
+   */
+  this._givenConfiguration = isPlainObject(givenConfig) ? givenConfig : null;
+  /**
+   * The _logger property holds a logger instance which, if valid, will be used
+   * to log any debug or error information generated by the instance.
+   * @memberof Configuration
+   * @private
+   * @type {Object|Null}
+   * @defaultvalue null
+   */
+  this._logger = isObject(logger) ? logger : null;
```

Comment:
@ofrobots  removing this since this is better fixed by the follow-on commit which introduces the logger a require